### PR TITLE
week2 - longest_substring

### PR DESCRIPTION
**Pull request by maknoon**
- TEBOW took 23min, implementation and testing took much longer (about 40min).
- `longest_substring.cc` - execute `make ls && ./ls test` to run the tests. we assume that the charset is limited to 128 (ASCII), and then use an array of int to store the index of char we are checking each time as we iterate thru the input string. this optimizes the solution to O(n)! what follows is that we can compare size of the substr within the defined limits and take the maximum `substr.size()` as the final output.
- `TODO`: this week's problem from the lecture as a refresher. also optimize it more - should add case if the end index is already going to be shorter than the current greatest substring

### CODE REVIEW

File: week2/longest_substring.cc [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                                       // str.substr(end,start) aka end-start+1
+     charset[str[end]] = end + 1; // update the charset of positions with the end + 1
+   }
+
+   // return length of the established greatest substring with nonrepeating chars
+   return max;
+ }
+
+/*
+ * for testing purposes
```
Comment (charientism) [inline week2/longest_substring.cc:61]: Thanks for testing/including your test cases!

File: week2/longest_substring.cc [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+   // we want to track our current greatest substring as we iterate
+   int start = 0; int end;
+   int max = 0;
+   // assume that the charset is limited size of ASCII charset
+   int charset[128] = {0};
+
+   // iterate thru the string accordingly to find the greatest substring
+   for (end = 0; end < str.size(); end++)
+   {
+#ifdef DEBUG
```
Comment (charientism) [inline week2/longest_substring.cc:45]: Don't forget to remove extra code :)

File: week2/longest_substring.cc [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+     std::string greatest_substr = str.substr(start,end);
```
Comment (charientism) [inline week2/longest_substring.cc]: Was this printing what you expected? If I remember correctly, substr's second arg is the length of the new substring, not the ending index

File: week2/longest_substring.cc [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+   // null exception
+   if (str == "") return -1;
+   // optimize. don't iterate if size is already 1
+   if (str.size() == 1) return 1;
+
+   // we want to track our current greatest substring as we iterate
+   int start = 0; int end;
+   int max = 0;
+   // assume that the charset is limited size of ASCII charset
+   int charset[128] = {0};
```
Comment (charientism) [inline week2/longest_substring.cc:40]: Good job! Storing the "next safe" index for each character is the optimized solution for this problem :)

File: week2/longest_substring.cc [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+//== FUNCTIONS
+/*
+ * int longest_substr(string str)
+ * input: str - the string to evaluate
+ * output: int - size of the greatest substring with no repeated chars of str
+ * notes: this is the main implementation
+ */
+ int longest_substr(std::string str)
+ {
+   // null exception
```
Comment (charientism) [inline week2/longest_substring.cc:31]: Thanks for commenting - all of the comments you've written in this method are great things to talk about aloud when you're going through the problem with an interviewer

**Review by charientism — APPROVED** [review]

File: week2/longest_substring.cc (hunk 3 above)
Comment (maknoon) [reply week2/longest_substring.cc]: oops yes, 2nd arg should be the strlen specifier. i had `end-start` there before, must have accidentally deleted it in the text editor by mistake

File: week2/longest_substring.cc (hunk 1 above)
Comment (maknoon) [reply week2/longest_substring.cc:61]: thanks for the code review! :)

**Pull request merged**